File: clients/android/app/src/main/java/network/buildit/modules/search/providers/SearchProvider.kt
package network.buildit.modules.search.providers

import network.buildit.modules.search.models.FacetDefinition
import network.buildit.modules.search.models.FormattedSearchResult
import network.buildit.modules.search.models.SearchDocument
import network.buildit.modules.search.models.SearchResult

/**
 * Interface that modules implement to provide searchable content.
 *
 * Each module (events, messaging, documents, wiki, etc.) can register a provider
 * that knows how to:
 * - Index entities from that module
 * - Define facets for filtering
 * - Format search results for display
 *
 * Usage:
 * ```kotlin
 * class EventsSearchProvider @Inject constructor(...) : SearchProvider {
 *     override val moduleType = "events"
 *
 *     override fun indexEntity(entity: Any, groupId: String): SearchDocument? {
 *         val event = entity as? Event ?: return null
 *         return SearchDocument(
 *             id = "events:${event.id}",
 *             moduleType = moduleType,
 *             entityId = event.id,
 *             groupId = groupId,
 *             title = event.title,
 *             content = event.description ?: "",
 *             ...
 *         )
 *     }
 *     ...
 * }
 * ```
 */
interface SearchProvider {
    /**
     * Unique identifier for this module (e.g., "events", "documents", "messaging").
     * Must match the module's identifier in BuildItModule.
     */
    val moduleType: String

    /**
     * Whether this provider is currently enabled.
     * Allows disabling search for specific modules without unregistering.
     */
    val isEnabled: Boolean
        get() = true

    /**
     * Boost factor for results from this module (default: 1.0).
     * Higher values make results from this module rank higher.
     */
    val boostFactor: Double
        get() = 1.0

    /**
     * Converts an entity from this module into a SearchDocument.
     *
     * @param entity The entity to index (type depends on module)
     * @param groupId The group this entity belongs to
     * @return SearchDocument if indexable, null if entity should be skipped
     */
    fun indexEntity(entity: Any, groupId: String): SearchDocument?

    /**
     * Returns the facet definitions for this module.
     * Facets allow users to filter search results by module-specific attributes.
     *
     * Example facets for events: date range, visibility level, has RSVP
     * Example facets for documents: document type, has attachments, last edited
     */
    fun getFacetDefinitions(): List<FacetDefinition>

    /**
     * Formats a search result for display in the UI.
     *
     * @param result The raw search result
     * @return Formatted result with title, subtitle, icon, navigation route, etc.
     */
    fun formatResult(result: SearchResult): FormattedSearchResult

    /**
     * Returns all indexable entities for a group.
     * Called during full reindex operations.
     *
     * @param groupId The group to get entities for
     * @return List of entities to index
     */
    suspend fun getIndexableEntities(groupId: String): List<Any>

    /**
     * Called when an entity is created or updated.
     * Providers can use this to trigger incremental index updates.
     *
     * @param entityId The entity that was modified
     * @param groupId The group the entity belongs to
     * @return The updated SearchDocument, or null if entity is not indexable
     */
    suspend fun onEntityChanged(entityId: String, groupId: String): SearchDocument? {
        // Default implementation - override for real-time indexing
        return null
    }

    /**
     * Called when an entity is deleted.
     *
     * @param entityId The entity that was deleted
     * @param groupId The group the entity belonged to
     */
    suspend fun onEntityDeleted(entityId: String, groupId: String) {
        // Default no-op - override to handle deletion
    }

    /**
     * Extracts tags from an entity for indexing.
     *
     * @param entity The entity to extract tags from
     * @return List of tag strings
     */
    fun extractTags(entity: Any): List<String> {
        return emptyList()
    }

    /**
     * Generates a preview excerpt from entity content.
     *
     * @param entity The entity to generate excerpt from
     * @param maxLength Maximum excerpt length
     * @return Short text excerpt
     */
    fun generateExcerpt(entity: Any, maxLength: Int = 200): String? {
        return null
    }
}

/**
 * Registry for search providers.
 * Modules register their providers here to participate in search.
 */
interface SearchProviderRegistry {
    /**
     * Registers a search provider for a module.
     */
    fun registerProvider(provider: SearchProvider)

    /**
     * Unregisters a search provider.
     */
    fun unregisterProvider(moduleType: String)

    /**
     * Gets all registered providers.
     */
    fun getAllProviders(): List<SearchProvider>

    /**
     * Gets provider for a specific module.
     */
    fun getProvider(moduleType: String): SearchProvider?

    /**
     * Gets all enabled providers.
     */
    fun getEnabledProviders(): List<SearchProvider>
}

File: clients/ios/BuildIt/Modules/Search/Providers/SearchProvider.swift
// SearchProvider.swift
// BuildIt - Decentralized Mesh Communication
//
// Protocol definition for module search providers.
// Each module can register a provider to make its content searchable.

import Foundation

// MARK: - SearchProvider Protocol

/// Protocol that modules implement to provide searchable content
public protocol SearchProvider: Sendable {
    /// The module type this provider handles
    var moduleType: String { get }

    /// The icon name for this module's results
    var moduleIcon: String { get }

    /// Convert an entity to a search document
    /// - Parameters:
    ///   - entity: The entity to index
    ///   - groupId: The group this entity belongs to
    /// - Returns: A SearchDocument if the entity can be indexed, nil otherwise
    func indexEntity(_ entity: Any, groupId: String) async -> SearchDocument?

    /// Get facet definitions for this module
    /// - Returns: Array of facet definitions for filtering
    func getFacetDefinitions() -> [FacetDefinition]

    /// Format a search result for display
    /// - Parameter result: The raw search result
    /// - Returns: A formatted result ready for UI display
    func formatResult(_ result: SearchResult) async -> FormattedSearchResult

    /// Get all indexable entities for a group
    /// - Parameter groupId: The group to get entities for
    /// - Returns: Array of entities to index
    func getIndexableEntities(groupId: String) async -> [Any]

    /// Get entities updated since a given timestamp
    /// - Parameters:
    ///   - timestamp: Unix timestamp in milliseconds
    ///   - groupId: The group to check
    /// - Returns: Array of entities updated since the timestamp
    func getEntitiesUpdatedSince(_ timestamp: Int64, groupId: String) async -> [Any]

    /// Handle entity deletion - return document ID to remove
    /// - Parameter entityId: The ID of the deleted entity
    /// - Returns: The search document ID to remove from index
    func getDocumentIdForEntity(_ entityId: String) -> String
}

// MARK: - Default Implementations

public extension SearchProvider {
    /// Default implementation returns document ID in standard format
    func getDocumentIdForEntity(_ entityId: String) -> String {
        "\(moduleType):\(entityId)"
    }

    /// Default implementation returns empty array for incremental updates
    func getEntitiesUpdatedSince(_ timestamp: Int64, groupId: String) async -> [Any] {
        []
    }
}

// MARK: - SearchProviderConfig

/// Configuration for a module's search provider
public struct SearchProviderConfig: Codable, Sendable {
    /// Module type this provider handles
    public let moduleType: String

    /// Whether this provider is active
    public var enabled: Bool

    /// Facets this module provides
    public let facetDefinitions: [FacetDefinition]

    /// Result boost factor for this module (default: 1.0)
    public var boost: Double

    public init(
        moduleType: String,
        enabled: Bool = true,
        facetDefinitions: [FacetDefinition] = [],
        boost: Double = 1.0
    ) {
        self.moduleType = moduleType
        self.enabled = enabled
        self.facetDefinitions = facetDefinitions
        self.boost = min(max(boost, 0), 10)
    }
}

// MARK: - SearchProviderRegistry

/// Registry for managing search providers
public actor SearchProviderRegistry {
    /// Singleton instance
    public static let shared = SearchProviderRegistry()

    /// Registered providers by module type
    private var providers: [String: any SearchProvider] = [:]

    /// Provider configurations
    private var configs: [String: SearchProviderConfig] = [:]

    private init() {}

    /// Register a search provider
    /// - Parameter provider: The provider to register
    public func register(_ provider: any SearchProvider) {
        providers[provider.moduleType] = provider
        if configs[provider.moduleType] == nil {
            configs[provider.moduleType] = SearchProviderConfig(
                moduleType: provider.moduleType,
                facetDefinitions: provider.getFacetDefinitions()
            )
        }
    }

    /// Unregister a provider
    /// - Parameter moduleType: The module type to unregister
    public func unregister(_ moduleType: String) {
        providers.removeValue(forKey: moduleType)
        configs.removeValue(forKey: moduleType)
    }

    /// Get a provider for a module type
    /// - Parameter moduleType: The module type
    /// - Returns: The provider if registered
    public func getProvider(for moduleType: String) -> (any SearchProvider)? {
        providers[moduleType]
    }

    /// Get all registered providers
    /// - Returns: Array of all providers
    public func getAllProviders() -> [any SearchProvider] {
        Array(providers.values)
    }

    /// Get all enabled providers
    /// - Returns: Array of enabled providers
    public func getEnabledProviders() -> [any SearchProvider] {
        providers.values.filter { provider in
            configs[provider.moduleType]?.enabled ?? true
        }
    }

    /// Get configuration for a provider
    /// - Parameter moduleType: The module type
    /// - Returns: The configuration if exists
    public func getConfig(for moduleType: String) -> SearchProviderConfig? {
        configs[moduleType]
    }

    /// Update provider configuration
    /// - Parameters:
    ///   - moduleType: The module type
    ///   - config: The new configuration
    public func updateConfig(_ moduleType: String, config: SearchProviderConfig) {
        configs[moduleType] = config
    }

    /// Enable or disable a provider
    /// - Parameters:
    ///   - moduleType: The module type
    ///   - enabled: Whether to enable or disable
    public func setEnabled(_ moduleType: String, enabled: Bool) {
        if var config = configs[moduleType] {
            config.enabled = enabled
            configs[moduleType] = config
        }
    }

    /// Get all facet definitions from all providers
    /// - Returns: Combined facet definitions
    public func getAllFacetDefinitions() -> [FacetDefinition] {
        providers.values.flatMap { $0.getFacetDefinitions() }
    }

    /// Get registered module types
    /// - Returns: Array of module type strings
    public func getRegisteredModuleTypes() -> [String] {
        Array(providers.keys).sorted()
    }
}
